File: julia/realtime.jl
# Real-time neural data filtering

using ArgParse
using Sockets
using HDF5
using Statistics
using Printf

# Byte used for "acknowledge" signal
const ACK_BYTE = 0x06

# Number of past signal values to use for filter
FILTER_ORDER = 5

# Learning rate for filter
FILTER_MU = 0.01

# Number of points in signal to read
const N_PTS = 1000


# Echo filter
mutable struct FilterAutoEcho
    FilterAutoEcho() = new()
end

function predict_next(flt::FilterAutoEcho, x::Array{Float64})::Array{Float64}
    return x
end


function hdr_encode(n_neurons::Integer)
    return reinterpret(UInt8, UInt16[convert(UInt16, n_neurons)])
end


function hdr_decode(hdr_bytes::Array{UInt8})
    return reinterpret(UInt16, hdr_bytes)[1]
end


function spks_encode(spks::Array{Int8})
    return reinterpret(UInt8, spks)
end


function spks_decode(spk_bytes::Array{UInt8})
    return reinterpret(Int8, spk_bytes)
end


function fpred_encode(fpred::Array{Float64})
    return reshape(reinterpret(UInt8, fpred), :, 1)
end


function fpred_decode(fpred_bytes::Array{UInt8})
    return reinterpret(Float64, reshape(fpred_bytes, 8, :))
end


function probe_mode(host::IPAddr, port::Integer, in_fpath::String, out_fpath::String)

    # Load test signal from file (need to swap axes; signal stored by NumPy)
    println("Loading data from '$in_fpath'...")
    spks = h5open(in_fpath, "r") do file
        permutedims(read(file, "spks"), [2, 1])
    end
    println("Done.")
    
    # Number of neurons (must be less than 65535)
    n_neurons = size(spks, 1)

    println("Connecting to $host:$port...")
    conn = connect(host, port)
    println("Done.")

    println("Sending header information...")

    # Send header
    hdr_bytes = hdr_encode(n_neurons)
    write(conn, hdr_bytes)
    
    # Receive ACK byte
    hdr_resp = read(conn, UInt8)
    if hdr_resp != ACK_BYTE
        error("Response to header not ACK")
    end

    println("Done.")

    println("Sending signal...")

    filter_preds = fill(NaN, n_neurons, N_PTS)
    rt_times_us = fill(NaN, N_PTS)

    for i in 1:N_PTS

        t_start_ns = time_ns()

        # Write spike counts (just neuron 1)
        write(conn, spks_encode(spks[:, i]))

        # Read filter value from client (8 bytes per neuron)
        filter_preds[:, i] = fpred_decode(read(conn, n_neurons * 8))

        # Compute round-trip latency (microseconds)
        rt_times_us[i] = (time_ns() - t_start_ns) / 1000

    end
    println("Done.")

    println("Closing...")
    close(conn)
    println("Done.")

    # Print average round-trip time
    mean_time = mean(rt_times_us)
    median_time = median(rt_times_us)
    @printf("Mean round-trip latency: %.2f μs\n", mean_time)
    @printf("Median round-trip latency: %.2f μs\n", median_time)

    # Write filter predictions and round-trip times to output file
    println("Writing data to '$out_fpath'...")
    h5open(out_fpath, "w") do file
        write(file, "filter_preds", filter_preds)
        write(file, "rt_times_us", rt_times_us)
    end
    println("Done")

end


function processor_mode(host::IPAddr, port::Integer, filter_type::String)

    # Connect to client
    println("Starting server at $host:$port...")
    server = listen(host, port)
    conn = accept(server)

    # TODO: Add address to printout
    println("Connected to probe.")

    println("Receiving header information...")

    # Receive header
    hdr_bytes = read(conn, 2)
    n_neurons = hdr_decode(hdr_bytes)
    
    # Send ACK
    write(conn, ACK_BYTE)

    println("Done.")

    println("Filtering signal...")

    if filter_type == "echo"
        flt = FilterAutoEcho()
    else
        error("Filter not implemented!")
    end

    while !eof(conn)

        # Receive spikes from probe (1 byte per neuron)
        spks = convert(Array{Float64}, spks_decode(read(conn, n_neurons)))

        # "Echo filter"
        fpred = predict_next(flt, spks)

        # Send filter prediction to probe
        #write(conn, fpred_encode(fpred))
        
        fpred_bytes = fpred_encode(fpred)

        t_start_ns = time_ns()

        write(conn, fpred_bytes)

        time = (time_ns() - t_start_ns) / 1000
        println("time (us):")
        println(time)


    end

    println("Done.")

    close(conn)

end


function main()

    settings = ArgParseSettings()

    @add_arg_table settings begin
        "probe"
            help = "probe mode"
            action = :command
        "processor"
            help = "processor mode"
            action = :command
    end

    @add_arg_table settings["probe"] begin
        "--host", "-a"
            arg_type = IPv4 
        "--port", "-p"
            arg_type = Int
        "--input", "-i"
            arg_type = String
        "--output", "-o"
            arg_type = String
    end

    @add_arg_table settings["processor"] begin
        "--host", "-a"
            arg_type = IPv4 
        "--port", "-p"
            arg_type = Int
        "--filter", "-f"
            arg_type = String
    end

    args = parse_args(settings)

    # TODO: Figure out how to parse IP address in ArgParse
    if args["%COMMAND%"] == "probe"
        cargs = args["probe"]
        probe_mode(cargs["host"], cargs["port"], cargs["input"], cargs["output"])
    elseif args["%COMMAND%"] == "processor"
        cargs = args["processor"]
        processor_mode(cargs["host"], cargs["port"], cargs["filter"])
    end
end

main()

File: c/realtime.c
/* Real-time neural data filtering using C

The current iteration of this code is meant as a prototype to test if C
is a suitable language to use for the production system. Because speed is an
important part of this system, this program measures the round-trip latency
of each set of spike counts it sends, and saves the times to the output file
for further analysis.

*/

#include <stdlib.h>
#include <stdio.h>
#include <string.h> 
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>


// TODO: Replace these eventually
#define HOST "127.0.0.1"
#define PORT 8889
#define N_NEURONS 5


// Probe mode
int probe_mode(char* host, int port) {

    printf("probe mode!\n");

    int spks[N_NEURONS] = {1, 2, 3, 4, 5};
    int filter_preds[N_NEURONS];

	int sock;
  	struct sockaddr_in server;

	// Create socket
  	sock = socket(AF_INET, SOCK_STREAM, 0);
  	if (sock < 0) {
    	perror("Cannot create socket");
    	exit(1);
  	}

    // Connect to server
  	server.sin_family = AF_INET;
  	server.sin_addr.s_addr = inet_addr(host);
  	server.sin_port = htons(port);
  	if (connect(sock, (struct sockaddr *)&server, sizeof(server)) < 0) {
    	perror("Cannot connect to server");
    	exit(2);
  	}

    // Send spike counts
    if (send(sock, &spks, N_NEURONS * sizeof(int), 0) < 0) {
        perror("Send failed");
        exit(3);
    }
  
    // Receive a reply from the server
    if (recv(sock, &filter_preds, N_NEURONS * sizeof(int), 0) < 0) {
        puts("recv failed");
        return 0;
    }
  
    // Print response
    printf("Server reply :\n");
    for (int i = 0; i < N_NEURONS; i++) {
        printf("%d\n", filter_preds[i]);
    }
  
    // close the socket
    close(sock);
    return 0;
}


// Processor mode
int processor_mode(char* host, int port) {

	int socket_desc, client_sock, c, read_size;
    struct sockaddr_in server, client;
    int message[N_NEURONS];
  
    // Create socket
    socket_desc = socket(AF_INET, SOCK_STREAM, 0);
    if (socket_desc == -1) {
        printf("Could not create socket");
    }
  
    // Prepare the sockaddr_in structure
    server.sin_family = AF_INET;
    server.sin_addr.s_addr = INADDR_ANY;
    server.sin_port = htons(port);
  
    // Bind the socket
    if (bind(socket_desc, (struct sockaddr*)&server, sizeof(server)) < 0) {
        perror("bind failed. Error");
        return 1;
    }
  
    // Listen to the socket
    listen(socket_desc, 3);
  
    puts("Waiting for incoming connections...");
    c = sizeof(struct sockaddr_in);
  
    // accept connection from an incoming client
    client_sock = accept(socket_desc, (struct sockaddr*)&client, (socklen_t*)&c);
  
    if (client_sock < 0) {
        perror("accept failed");
        return 1;
    }
  
    puts("Connection accepted");
  
    // Receive a message from client
    while ((read_size = recv(client_sock, &message, N_NEURONS * sizeof(int), 0)) > 0) {
  
        write(client_sock, &message, N_NEURONS * sizeof(int));
    }
  
    if (read_size == 0) {
        puts("Client disconnected");
    }
    else if (read_size == -1) {
        perror("recv failed");
    }
 	return 0;
} 


int processor_mode_old(char* host, int port) {

    printf("processor mode!\n");

	/* Echo server code */

	int serverFd, clientFd;
  	struct sockaddr_in server, client;
  	int len;
  	//int port = 1234;
  	char buffer[1024];
  	//if (argc == 2) {
    //	port = atoi(argv[1]);
  	//}
  	serverFd = socket(AF_INET, SOCK_STREAM, 0);
  	if (serverFd < 0) {
    	perror("Cannot create socket");
    	exit(1);
  	}
  	server.sin_family = AF_INET;
  	server.sin_addr.s_addr = INADDR_ANY;
  	server.sin_port = htons(port);
  	len = sizeof(server);
  	if (bind(serverFd, (struct sockaddr *)&server, len) < 0) {
    	perror("Cannot bind sokcet");
    	exit(2);
  	}
  	if (listen(serverFd, 10) < 0) {
    	perror("Listen error");
    	exit(3);
  	}
  	while (1) {
    	len = sizeof(client);
    	printf("waiting for clients\n");
    	if ((clientFd = accept(serverFd, (struct sockaddr *)&client, &len)) < 0) {
      		perror("accept error");
      		exit(4);
    	}
    	char *client_ip = inet_ntoa(client.sin_addr);
    	printf("Accepted new connection from a client %s:%d\n", client_ip, ntohs(client.sin_port));
    	memset(buffer, 0, sizeof(buffer));
    	int size = read(clientFd, buffer, sizeof(buffer));
    	if ( size < 0 ) {
      		perror("read error");
      		exit(5);
    	}
    	printf("received %s from client\n", buffer);
    	if (write(clientFd, buffer, size) < 0) {
      		perror("write error");
      		exit(6);
    	}
    	close(clientFd);
  	}

  close(serverFd);
  return 0;

}


// Print usage message
void print_usage() {

    printf("Usage: realtime [probe, processor]\n");

}

int main(int argc, char **argv) {

    // Parse mode ('probe' or 'processor') from argument list
    if (argc < 2) {
        print_usage();
        return 0;
    } 
    else {
        if (strcmp(argv[1], "probe") == 0) {
            return probe_mode(HOST, PORT);
        }
        else if (strcmp(argv[1], "processor") == 0) {
            return processor_mode(HOST, PORT);
        }
        else {
            print_usage();
            return 0;
        }
    }

    return 0;
}
